RIP-2 Storage Layout

FintanH opened the pull request:
Add RIP-2 describing the layout of the git storage for the Heartwood protocol.

keepsimple1 commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,518 @@
+
+If the `RAD_HOME` environment variable is set, the directory it points
+to will be used as the base directory. If `RAD_HOME` is set it will
+take priority as the base directory.
```
Is the last sentence repeating the previous one?

FintanH replied:
Oh yes, thank you :bow:

keepsimple1 commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,518 @@
+
+| Value                        | Example                       |
+| ---------------------------- | ------------------------------|
+| `$RAD_HOME/.radicle/storage` | /home/alice/MyRadicle/storage |
```
What is the value of `$RAD_HOME` in the above? I don't understand how the full path ends up not in `.radicle`?

FintanH replied:
It's an environment variable, so the user can set it to whatever they want. But actually, it shouldn't end up in `.radicle` :) It should in fact be like this:
```suggestion
| `$RAD_HOME/storage` | /home/alice/MyRadicle/storage |
```

keepsimple1 commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,518 @@
+here onwards.
+
+This means that a peer's references will be scoped by
+`refs/namespaces/<pubkey>`. We demonstrate this organisation below:
```
Does this rule apply to the local user / node?  I.e. do the `peers` include the local node as well?

FintanH replied:
I point this out above https://github.com/radicle-dev/rips/pull/1/files#diff-0ccf2baf8bebd800e0e6c8723c5dba6b925263f9e64b48643a619941fc35f4cdR120

keepsimple1 commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,518 @@
+
+In the next sections we will cover how the above layout works.
+
+### Remote Replication
```
nit:  it seems there are no `##` headers before this line?

FintanH replied:
I followed suit from the first RIP where it uses a series of `-` instead of `##`

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+
+One of the key components of the Radicle network is the storage
+layer. A key aspect of this component is that it is local-first. It
+must be able to hold the view of the local operator's view of a
```
```suggestion
must be able to hold the local operator's view of a
```

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+
+With the above in mind, this RIP proposes a storage layer that
+fulfills the following requirements:
+1. It can be stored, locally, on disk.
```
I'd rephrase this to carry more meaning, into something like "It is able to keep a local copy of the entire working dataset"

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+* [Copyright](#copyright)
+
+Storage Home
+------------
```
Do you think it might be better to avoid specifying RAD_HOME entirely here? In which case we could say something like `let '/storage' be a directory under which all radicle git state shall be stored`...

Because it seems like a bit of a distraction when what we're interested in is more what is *in* the storage folder.

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+i.e. `storage/<resource>/refs/{heads, tags}`. These namespaces must be
+reserved for canoncial references -- references that are agreed,
+collaboratively, as published and stable. How canonical references are
+decided and written is left for further RIP.
```
Actually there is one thing we could say here, it's that these references are *not* replicated. They are set locally, according to some heuristics that are TBD.

FintanH replied:
But _should_ we say that here? That feels like it should be defined in a RIP that tackles replication directly, IMO.

cloudhead replied:
Hmmm, think you're right.

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+1. Remote replication
+2. Linking to a working copy
+
+In the next sections we will cover how the above layout works.
```
```suggestion
In the next sections we will cover how the above works with the storage layout works.
```
Makes more sense to me?

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+working copy corresponds to one resource in the storage -- where the
+operator makes their own changes to the code base. This can be
+compared to the way one would `git clone` from a mirrored repository,
+e.g. GitHub, GitLab, etc. They can then make changes and push to the
```
```suggestion
e.g. GitHub, GitLab, etc. One can then make changes and push to the
```

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+The connection between the working copy and the storage is maintained
+by a series of `git` [remotes][git-remotes]. Each `git` remote
+represents a single peer -- or namespace -- for that resource.
+
```
I'd add here something like:
> One remote in particular is always setup when a working copy is created: the `rad` remote, which allows operators to push to their own copy of the resource.

FintanH replied:
I would argue that that's a decision made by the application, i.e. there isn't anything special about a `rad` remote, they could name it `local`, `origin`, or even `foobar`. But I can add a suggestion for doing so to make it easier to use.

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+signing pushed references to one's own namespace. This is discussed in
+[Future Work](#Future-Work).
+
+To ensure `gitnamespaces` worked as expected the author ran through a
```
```suggestion
To ensure `gitnamespaces` work as expected the author ran through a
```

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+These three requirements can be solved by introducing a
+`git-remote-rad` helper binary that can supply the namespace, ensure
+the peer namespaces are respected, and that the signed references are
+updated. This is left to another RIP to define.
```
I'd actually say that we should move this little section up, as a solution to the above problems around verifying the push. We don't need to really specify the remote-helper that much; but I think this RIP is the place to do so. I'd talk about:
1. The `rad://` scheme
2. The handling of namespaces (ie. a user should not be able to push to a namespace other than his own)

cloudhead reviewed (commented):
This is looking good, but I'm wondering if the scope is too narrow. There's two things I'd consider:

1. We should probably define now the formats of the user keys and project ids. I think this should probably go in RIP 1. This would allow this RIP to be much more specific about the storage layout, which I think will be really helpful to the reader.
2. I'd consider talking about signed-refs here, and more generally about `rad/*` special refs; because it is part of the storage layout. This might mean that this should be RIP 3, and RIP 2 should be abouy identity. I'm just trying to make this RIP more useful to implementors, and right now I think we're limiting ourselves too much due to unspecified stuff.

What do you think?

keepsimple1 commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+The layout must support multiple resources and multiple peers per
+resource. Under the `storage` directory, each resource will be a [bare
+git repository][git-bare]. For each resource in the storage, it must
+have a stable and unique identifier. This is to ensure that resources
```
Is this `unique identifier` globally unique, or only unique on this node?

FintanH replied:
Aye, it is. I'll clarify it in the doc.

slack-coder commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+interested in. For local users this improves the experience of
+interacting with the data -- not needing a server to compute data for
+you and being able to work offline. For local users and seeding nodes
+in the network this improves the availability of data.
```
Its also about censorship resistance

slack-coder commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+`storage` found under one of two base directories.
+
+If the `RAD_HOME` environment variable is set, the directory it points
+to will be used as the base directory.
```
Maybe introducing what's being effected by RAD_HOME first will help readers?

slack-coder replied:
for example, "RAD's storage is located by default under the user directory at X.   It can be overridden using the RAD_HOME envirnment variable."

It builds up to it using only defined concepts.

FintanH replied:
> Maybe introducing what's being effected by RAD_HOME first will help readers?

The section is under "Storage Home", do you think that that is not clear enough? :thinking: 

I also finished the last paragraph with:
> The storage must be stored under the name `storage` found under one of two base directories.

But maybe I can say that is either the user's home or a preset `RAD_HOME`.

slack-coder commented on 0002-storage-layout.md:
```diff
@@ -0,0 +1,517 @@
+License: CC0-1.0
+---
+
+RIP #2: Storage Layout
```
Would other parties making a 'RAD' be expected to use this storage format?

cloudhead replied:
what's a RAD?

FintanH replied:
If I'm understanding you correctly, the idea of a RIP is that these are expected implementations of any kind of Radicle clone. So if someone wrote a protocol and client in Go, they would be expected to do so following the RIPs to have parity with the Rust implementation.

slack-coder replied:
Thanks @FintanH.  Though the design may be good, would it be better to leave this out?  It looks to be an internal implementation detail.  Third party implementations would be considered a success right?  Minimizing RIPs to the protocol would reduce friction.

FintanH replied:
Why do you think the storage layout is an implementation detail?

slack-coder replied:
To be clear, its a question of Radicle Improvement Proposal's  over the design discussion.  Its good to have the design discussion for Heartwood in either case.

What are RIPs for?  They come across to me as design proposals for Radicle's Heartwood project.  Take defining `RAD_HOME` for example.  I recommend minimizing them to what's expected for interoperability for third party RAD applications.

RIPs appear to take from BIPs, and we missed an important aspect.  BIPs have three purposes as shown below.  Maybe we should adopt 'Process' as a use for RIPs?  The [first BIP](https://github.com/bitcoin/bips/blob/master/bip-0001.mediawiki) is about introducing BIPs, defining process and what they are for.

```
Standards Track BIPs – Such types of BIPs entail making changes to the [network](https://en.bitcoinwiki.org/wiki/Network) protocol, [block](https://en.bitcoinwiki.org/wiki/Block), or [transaction](https://en.bitcoinwiki.org/wiki/Transaction) validation method. It also intends to affect the interoperability of the two versions of BIPs or Bitcoin. This type of BIP certainly requires community consensus. An example of this is BIP 91.
    Informational BIPs – Such types of BIPs highlight the design issues, general guidelines, and supporting information. Informational BIPs, as the name suggests, are just for information’s sake and can be taken seriously or ignored by the community. An example of this is BIP 32.
    Process BIPs – These types of BIPs describe or propose a change in the process. They are similar to Standards Track BIPs and require community consensus. They can’t be ignored, but unlike Standards Track BIPs, they intend to be applied outside the [Bitcoin](https://en.bitcoinwiki.org/wiki/Bitcoin) protocol. An example of this is BIP 2.
```

slack-coder replied:
> Why do you think the storage layout is an implementation detail?

The RAD_HOME environment variable, and the directory layout of its contents.  My perspective being what's needed for different users to interact with each other using Radicle, as opposed to different applications accessing Radicle project information for the same user.

slack-coder replied:
> To be clear, its a question of Radicle Improvement Proposal's over the design discussion. Its good to have the design discussion for Heartwood in either case.
> 
> What are RIPs for? They come across to me as design proposals for Radicle's Heartwood project. Take defining `RAD_HOME` for example. I recommend minimizing them to what's expected for interoperability for third party RAD applications.
> 
> RIPs appear to take from BIPs, and we missed an important aspect. BIPs have three purposes as shown below. Maybe we should adopt 'Process' as a use for RIPs? The [first BIP](https://github.com/bitcoin/bips/blob/master/bip-0001.mediawiki) is about introducing BIPs, defining process and what they are for.
> 
>     * Standards Track BIPs – Such types of BIPs entail making changes to the [network](https://en.bitcoinwiki.org/wiki/Network) protocol, [block](https://en.bitcoinwiki.org/wiki/Block), or [transaction](https://en.bitcoinwiki.org/wiki/Transaction) validation method. It also intends to affect the interoperability of the two versions of BIPs or Bitcoin. This type of BIP certainly requires community consensus. An example of this is BIP 91.
> 
>     * Informational BIPs – Such types of BIPs highlight the design issues, general guidelines, and supporting information. Informational BIPs, as the name suggests, are just for information’s sake and can be taken seriously or ignored by the community. An example of this is BIP 32.
> 
>     * Process BIPs – These types of BIPs describe or propose a change in the process. They are similar to Standards Track BIPs and require community consensus. They can’t be ignored, but unlike Standards Track BIPs, they intend to be applied outside the [Bitcoin](https://en.bitcoinwiki.org/wiki/Bitcoin) protocol. An example of this is BIP 2.

To keep the topic to this RIP, l've created [an issue](https://github.com/radicle-dev/rips/issues/3) for fork this conversation to.

FintanH replied:
I've removed mention of RAD_HOME and the storage home in general :) https://github.com/radicle-dev/rips/pull/1/commits/d301f209d688e45ee6d5da7fd1d6daab6b865ebf

FintanH commented:
Made a few changes in a fixup commit. Ready for another round of review :)

cloudhead commented on 0002-storage-layout.md:
```diff
@@ -204,20 +206,20 @@ A working copy is a local replica of the resource -- that is each
-The `url` must be able to resolve the local storage's resource that
-this working copy corresponds to.
+The `url` of the remote must be able to resolve the local storage's
+resource that this working copy corresponds to.
```
"corresponding to this working copy" reads nicer I think

FintanH commented:
Some more updates:
* Changed it to RIP#3 and rebased off of RIP#2 PR
* Edited links to mirror RIP#2 style
* Mentioned identifiers as established in RIP#2.
* Some wording fix ups.

gsaslis commented on 0003-storage-layout.md:
````diff
@@ -0,0 +1,564 @@
+This will do what you expect when running:
+
+```
+$ get fetch alice
````
```suggestion
$ git fetch alice
```

... `get fetch` would be confusing in all sorts of weird ways, not to mention I'd never be able to remember if I `get fetch ...` or `fetch get ...`  ;p

gsaslis commented on 0003-storage-layout.md:
```diff
@@ -0,0 +1,564 @@
+  the storage.
+* The `<pubkey>` is the namespace for which the the `--namespace`
+  option will be set to. If `<pubkey>` is not specified, then the
+  local operator's key is used.
```
This whole section about how the `rad://` scheme works is very well explained, thanks for that!

gsaslis commented on 0003-storage-layout.md:
```diff
@@ -0,0 +1,564 @@
+
+* The `rad://` scheme informs `git` to look for an executable
+  `git-remote-rad` which will be executed during a `git-push` or
+  `git-fetch`.
```
maybe a quick note could be added about where `git` would look for `git-remote-rad` ?

FintanH replied:
Ya, that's a good shout :+1:

gsaslis commented on 0003-storage-layout.md:
```diff
@@ -0,0 +1,564 @@
+
+#### Authorization
+
+Since pushing to other peer's namespaces is disallowed, the value of
```
"disallowed" by whom? the Remote Helper? 

if so, perhaps we could reword this to sth like: 

> Since the Remote Helper MUST disallow pushing to other peer's namespaces ...

FintanH replied:
Ya, essentially it's a protocol invariant. I'll make a note of this :)

gsaslis commented on 0003-storage-layout.md:
```diff
@@ -0,0 +1,564 @@
+#### Signing the Reference Set
+
+To ensure that the local operator's reference set is cryptographically
+verifiable, the reference set is signed with the operator's key during
```
it is a little unclear to me who does this signing ... I know we are in the remote helper section, but if it is the remote helper that takes care of this can we explicitly mention that here?

gsaslis commented on 0003-storage-layout.md:
```diff
@@ -0,0 +1,564 @@
+Appendix
+--------
+
+### Worked Example
```
Perhaps it is not within scope of this RIP, but it occurred to me that the example doesn't seem to cover signed commits. Considering there is earlier talk about signing happening during `git push` could we extend the example to cover that? (and if there is some provision to disallow pushing to other peer's namespaces it'd also be great if we showed how that happens!)

FintanH replied:
Commit signing is out of scope. The protocol isn't concerned with signing individual commits, but if you do sign your commits then nothing here changes :)

gsaslis reviewed (commented):
@FintanH ok, so I don't know if this is the proper place to provide feedback, but I had some questions along the way as I was reading the doc to understand what's coming and I thought I'd drop some notes. 

It makes sense overall - I've just suggested some clarifications to be added and perhaps some wording improvements. 

Thanks for putting this together!

FintanH commented on 0003-storage-layout.md:
```diff
@@ -0,0 +1,564 @@
+------------
+
+It is trivial to store `git` repositories, locally, on disk. It is
+also efficient to do so. The storage must be stored under the name
```
Aye, I believe that @cloudhead isn't a fan of that approach though :)

FintanH commented on 0003-storage-layout.md:
````diff
@@ -0,0 +1,564 @@
+
+```
+storage/
+  <RID>/   # Some resource, e.g. a project
````
Please refer to [Zooko's triangle](https://en.wikipedia.org/wiki/Zooko%27s_triangle) :)

gsaslis commented:
@FintanH another question: 

does Heartwood still (plan to) include "Collaborative Objects (Cobs)" ? If so, perhaps it would make sense to show how these are stored in this RIP ?

gsaslis commented:
> They are being used :) There's already existing code for them in Heartwood. That being said, I think that they should be defined in a separate RIP and how they are stored would follow the rules set out in this RIP, i.e. separate category under a specific peer, for each peer.

ok, ok, thanks! 👍🏼
